hnn_core/extracellular.py

```diff
@@ -0,0 +1,601 @@
+                self._nrn_r_transfer.setrow(row,
+                                            h.Vector(segment_counts.sum(), 1.))
+
+        # record time for each array
+        self._nrn_times = h.Vector().record(h._ref_t)
+
+        # contributions of all segments on this rank to total calculated
+        # potential at electrode (_PC.allreduce called in _simulate_dipole)
+        self._nrn_voltages = h.Vector()
+        self._nrn_imem_vec = h.Vector(n_total_segments)
+
+        # Attach a callback for calculating the potentials at each time step.
+        # Enables fast calculation of transmembrane current (nA) at each
+        # segment. Note that this will run on each rank, so it is safe to use
+        # the extra_scatter_gather-method, which docs say doesn't support
+        # "multiple threads".
+        cvode.extra_scatter_gather(0, self._calc_potentials_callback)
+
+    @property
+    def _nrn_n_samples(self):
+        """Return the length (in samples) of the extracellular data."""
+        if self._nrn_voltages.size() % self.n_contacts != 0:
+            raise RuntimeError('Something went wrong: have {self.n_contacts}'
+                               f', but {self._nrn_voltages.size()} samples')
+        return int(self._nrn_voltages.size() / self.n_contacts)
+
+    def _get_nrn_voltages(self):
+        """The extracellular data (n_contacts x n_samples)."""
+        if len(self._nrn_voltages) > 0:
+            assert (self._nrn_voltages.size() ==
+                    self.n_contacts * self._nrn_n_samples)
+
+            # first reshape to a Neuron Matrix object
+            extmat = h.Matrix(self.n_contacts, self._nrn_n_samples)
+            extmat.from_vector(self._nrn_voltages)
+
+            # then unpack into 2D python list and return
+            return [extmat.getrow(ii).to_python() for
+                    ii in range(extmat.nrow())]
+        else:
+            raise RuntimeError('Simulation not yet run!')
+
+    def _get_nrn_times(self):
+        """The sampling time points."""
+        if self._nrn_times.size() > 0:
+            # NB _nrn_times is one sample longer than _nrn_voltages
+            return self._nrn_times.to_python()[:self._nrn_n_samples]
+        else:
+            raise RuntimeError('Simulation not yet run!')
+
+    def _calc_potentials_callback(self):
+        # keep all data in Neuron objects for efficiency
+
+        # 'gather' the values of seg.i_membrane_ into self.imem_vec
+        self._nrn_imem_ptrvec.gather(self._nrn_imem_vec)
+
+        # Calculate potentials by multiplying the _nrn_imem_vec by the matrix
+        # _nrn_r_transfer. This is equivalent to a row-by-row dot-product:
+        # V_i = SUM_j (R_i,j x I_j)
+        self._nrn_voltages.append(
+            self._nrn_r_transfer.mulv(self._nrn_imem_vec))
+        # NB all values appended to the h.Vector _nrn_voltages at current time
+        # step. The vector will have sixe (n_contacts x n_samples, 1), which
```

Comment:
what is `n_samples` ?

Reply:
bookmarked. To continue review from here tomorrow

Reply:
> what is `n_samples` ?

The number of times `_calc_potentials_callback` is called. It turns out to be `n_times` - 1 (see above...)